### PR TITLE
PP-1484 fix invalid reference handling in worldpay refunds

### PR DESCRIPTION
**Pull request by kakumara**
## WHAT

 - Changed the approach slightly to use GOVUK Pay external ID instead of a random UUID
 - So for worldpay our integration partners can cross reference with the same exertnal ID we present to them in Worlpay console.
 - For smartpay also we are using the same apporach, however we must store smartpay's pspReference as the link with gateway.
 - Removed the generateReference()  method which was in PaymentProvider generic interface which was not doing much for some gateways.
 - Fixed invalid tests and added gateway specific tests.

### CODE REVIEW

File: src/main/java/uk/gov/pay/connector/service/ChargeRefundService.java [hunk 1]
```diff
@@ -183,11 +181,21 @@ private void checkIfChargeIsRefundableOrTerminate(ChargeEntity reloadedCharge, E
         }
     }
 
+    /**
+     * <p>Smartpay -> We get the pspReference returned by them</p>
+     * <p>Worldpay -> Worldpay doesn't return reference. We use our externalId as thats what we sent in the request as our reference.</p>
+     * @see RefundGatewayRequest valueOf()
+     */
     private String getRefundReference(RefundEntity refundEntity, GatewayResponse<BaseRefundResponse> gatewayResponse) {
```
Comment (PauloPortugal) [inline src/main/java/uk/gov/pay/connector/service/ChargeRefundService.java]: @heathd & @ckalista this is the method where we might have a bit of contention and the reason we have this PR marked as REVIEW ONLY.
`return gatewayResponse.getBaseResponse().get().getReference().orElse(refundEntity.getExternalId());`

This line will work for WorldPay (`getExternalId()`) and Smartpay (`getReference()`. 
If the response is not successful, then we will return an empty string as a refund reference, this being a cue that something went wrong when refunding.

**Comment by ckalista** [comment]
Code review completed. It all looks good!

**Comment by ckalista** [comment]
Confirmed the existence of the bug by creating a payment in test environment for Worldpay. After refund, the notification sent by the payment provider could not be reconciled:
```
postman --authkey=xxx --integration ar                                                                          [14:19:35]
postman       | [INFO] Querying https://publicapi.test.pymnt.uk/v1/payments
postman       | [INFO] Querying https://publicapi.test.pymnt.uk/v1/payments/lvsdh7k65crbsd93o5501dnf82/refunds
{
  "payment_id": "lvsdh7k65crbsd93o5501dnf82",
  "_links": {
    "self": {
      "href": "https://publicapi.test.pymnt.uk/v1/payments/lvsdh7k65crbsd93o5501dnf82/refunds"
    },
    "payment": {
      "href": "https://publicapi.test.pymnt.uk/v1/payments/lvsdh7k65crbsd93o5501dnf82"
    }
  },
  "_embedded": {
    "refunds": [
      {
        "amount": 12345,
        "created_date": "2016-12-21T14:08:17.128Z",
        "refund_id": "d4k15q9cegcqpsv1mru0h69n60",
        "status": "submitted",
        "_links": {
          "self": {
            "href": "https://publicapi.test.pymnt.uk/v1/payments/lvsdh7k65crbsd93o5501dnf82/refunds/d4k15q9cegcqpsv1mru0h69n60"
          },
          "payment": {
            "href": "https://publicapi.test.pymnt.uk/v1/payments/lvsdh7k65crbsd93o5501dnf82"
          }
        }
      }
    ]
  }
}
```

**Comment by simad** [comment]
:+1: LGTM - merge on behalf of Paulo, who, I am sure, would surely approve 💯

**Pull request merged**